src/stream.rs

```diff
@@ -562,204 +563,166 @@ impl<T1: StreamPack> StreamPack for (T1,) {
     }
 }
 
-impl<T1: StreamPack, T2: StreamPack> StreamPack for (T1, T2) {
-    type StreamAvailableBundle = (T1::StreamAvailableBundle, T2::StreamAvailableBundle);
-    type StreamFilter = (T1::StreamFilter, T2::StreamFilter);
-    type StreamStorageBundle = (T1::StreamStorageBundle, T2::StreamStorageBundle);
-    type StreamInputPack = (T1::StreamInputPack, T2::StreamInputPack);
-    type StreamOutputPack = (T1::StreamOutputPack, T2::StreamOutputPack);
-    type Receiver = (T1::Receiver, T2::Receiver);
-    type Channel = (T1::Channel, T2::Channel);
-    type Buffer = (T1::Buffer, T2::Buffer);
-
-    fn spawn_scope_streams(
-        in_scope: Entity,
-        out_scope: Entity,
-        commands: &mut Commands,
-    ) -> (
-        Self::StreamInputPack,
-        Self::StreamOutputPack,
-    ) {
-        let t1 = T1::spawn_scope_streams(in_scope, out_scope, commands);
-        let t2 = T2::spawn_scope_streams(in_scope, out_scope, commands);
-        ((t1.0, t2.0), (t1.1, t2.1))
-    }
+macro_rules! impl_streampack_for_tuple {
+    ($($T:ident),*) => {
+        #[allow(non_snake_case)]
+        impl<$($T: StreamPack),*> StreamPack for ($($T,)*) {
+            type StreamAvailableBundle = ($($T::StreamAvailableBundle,)*);
+            type StreamFilter = ($($T::StreamFilter,)*);
+            type StreamStorageBundle = ($($T::StreamStorageBundle,)*);
+            type StreamInputPack = ($($T::StreamInputPack,)*);
+            type StreamOutputPack = ($($T::StreamOutputPack,)*);
+            type Receiver = ($($T::Receiver,)*);
+            type Channel = ($($T::Channel,)*);
+            type Buffer = ($($T::Buffer,)*);
+
+            fn spawn_scope_streams(
+                in_scope: Entity,
+                out_scope: Entity,
+                commands: &mut Commands,
+            ) -> (
+                Self::StreamInputPack,
+                Self::StreamOutputPack,
+            ) {
+                let ($($T,)*) = (
+                    $(
+                        $T::spawn_scope_streams(in_scope, out_scope, commands),
+                    )*
+                );
+                // Now unpack the tuples
+                (
+                    (
+                        $(
+                            $T.0,
+                        )*
+                    ),
+                    (
+                        $(
+                            $T.1,
+                        )*
+                    )
+                )
```

Comment:
This is a bit of a symbol soup that I didn't know how to format better, it takes a lot of lines but it would be pretty much unreadable if it was compressed in a single line